src/main/java/org/broadinstitute/hellbender/utils/iterators/IntervalAlignmentContextIterator.java

```diff
@@ -0,0 +1,109 @@
+ *
+ * In other words, this will iterate over each base of each interval and emit an alignment context.  If no reads overlap,
+ *  it will emit an empty alignment context.  Each empty alignment context will be a new instance.
+ */
+public class IntervalAlignmentContextIterator implements Iterable<AlignmentContext>, Iterator<AlignmentContext> {
+    private Iterator<AlignmentContext> alignmentContextIterator;
+    private IntervalLocusIterator intervalLocusIterator;
+    private SimpleInterval currentInterval;
+    private AlignmentContext currentAlignmentContext;
+    private SAMSequenceDictionary dictionary;
+
+    public IntervalAlignmentContextIterator(Iterator<AlignmentContext> alignmentContextIterator, IntervalLocusIterator intervalLocusIterator, SAMSequenceDictionary dictionary) {
+        this.alignmentContextIterator = alignmentContextIterator;
+        this.intervalLocusIterator = intervalLocusIterator;
+        this.dictionary = dictionary;
+
+        // Iterate once.
+        advanceIntervalLocus();
+        advanceAlignmentContext();
+        advanceAlignmentContextToCurrentInterval(this.currentInterval);
+    }
+
+    @Override
+    public Iterator<AlignmentContext> iterator() {
+        return this;
+    }
+
+    @Override
+    public boolean hasNext() {
+        return (intervalLocusIterator.hasNext()) || (currentInterval != null);
+    }
+
+    @Override
+    public AlignmentContext next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
+        }
+
+        final boolean isOverlaps = currentInterval.overlaps(currentAlignmentContext);
+        AlignmentContext result;
+
+        // Get ready to return result and determine what should go the next time next() is called.
+        if (isOverlaps) {
+            result = currentAlignmentContext;
+            advanceIntervalLocus();
+            advanceAlignmentContextToCurrentInterval(currentInterval);
+        } else {
+            result = createEmptyAlignmentContext(currentInterval);
+            advanceIntervalLocus();
+
+            if (currentInterval != null) {
+                final int comparison = IntervalUtils.compareLocatables(currentInterval, currentAlignmentContext, dictionary);
+                // Comparison should not be zero and if less than zero, we'd just advance the intervalLocus iterator,
+                //  which we are already doing.
+                //  Interval is after the next alignment context when comparison is greater than zero.
+                if (comparison > 0) {
+                    advanceAlignmentContextToCurrentInterval(currentInterval);
+                }
+            }
+        }
+
+        return result;
+
+    }
+
+    private AlignmentContext createEmptyAlignmentContext(final SimpleInterval interval) {
+        return new AlignmentContext(interval, new ReadPileup(interval, new ArrayList<>()));
+    }
+
+    private void advanceAlignmentContextToCurrentInterval(final SimpleInterval interval) {
```

Comment:
Thought it would be more flexible, but then it had bugs.  I've refactored as you suggest.

Done.